driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -264,23 +264,29 @@ public boolean isClosed() {
     }
     private void sendCommandMessage(final CommandMessage message,
                                     final ByteBufferBsonOutput bsonOutput, final SessionContext sessionContext) {
-        try {
-            if (sendCompressor == null || SECURITY_SENSITIVE_COMMANDS.contains(message.getCommandDocument(bsonOutput).getFirstKey())) {
+        if (sendCompressor == null || SECURITY_SENSITIVE_COMMANDS.contains(message.getCommandDocument(bsonOutput).getFirstKey())) {
+            try {
                 sendMessage(bsonOutput.getByteBuffers(), message.getId());
-            } else {
-                CompressedMessage compressedMessage = new CompressedMessage(message.getOpCode(), bsonOutput.getByteBuffers(),
-                        sendCompressor,
+            } finally {
+                bsonOutput.close();
+            }
+        } else {
+            List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
```

Comment:
The underlying issues is that #getByteBuffers duplicates the buffers, which requires that they be released independently of the BsonOutput that they came from.

Reply:
So compression causes a double copy of buffers?

Reply:
Two buffers, but different contents.  The first is the uncompressed message.  The second is the compressed message.